src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -1440,42 +1440,53 @@ MarkRanges MergeTreeDataSelectExecutor::filterMarksUsingIndex(
 
             if (index_helper->isVectorSimilarityIndex())
             {
-                /// An array of indices of useful ranges.
-                auto result = condition->getUsefulRanges(granule);
+                auto rows = condition->calculateApproximateNearestNeighbors(granule);
 
-                for (auto range : result)
+                std::sort(rows.begin(), rows.end());
+                rows.erase(std::unique(rows.begin(), rows.end()), rows.end()); /// duplicates are in theory not possible but who knows ...
+
+                const MergeTreeIndexGranularity & merge_tree_index_granularity = part->index_granularity;
+
+                for (auto row : rows)
                 {
-                    /// The range for the corresponding index.
+                    size_t num_marks = merge_tree_index_granularity.countMarksForRows(index_mark * index_granularity, row);
+                    size_t result_mark = (index_mark * index_granularity) + num_marks;
+
                     MarkRange data_range(
```

Comment:
That's actually a strange piece of code. if data_range.begin == date_range.end then getNumberOfMarks() == 0. otherwise, we have data_range = result_mark

Reply:
You mean why we we are using std::min and std::max in l. 1450 and 1451? To be honest, that is copypasta from l. 1472 and it is also not really clear to me why this is needed. It looks like it was added later to handle some weird edge case. I'd rather keep the logic for now ...